From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Oliver Scott <olivercscott@gmail.com>
Date: Mon, 25 Oct 2021 16:16:22 -0400
Subject: [PATCH 1/3] Wi-Fi timeout feature

Listen to changes to the wifi timeout setting as well as the network
and wifi state. Schedule an alarm to turn wifi off if it is enabled
but not connected

Co-authored-by: Tommy Webb <tommy@calyxinstitute.org>
Change-Id: I905a6ccdbd26c0d472e2bb9d03159fda430981c6
Signed-off-by: Pranav Vashi <neobuddy89@gmail.com>
---
 .../android/server/wifi/WifiServiceImpl.java  | 47 +++++++++++++++++++
 1 file changed, 47 insertions(+)

diff --git a/service/java/com/android/server/wifi/WifiServiceImpl.java b/service/java/com/android/server/wifi/WifiServiceImpl.java
index 72f8aa7..62a70d0 100644
--- a/service/java/com/android/server/wifi/WifiServiceImpl.java
+++ b/service/java/com/android/server/wifi/WifiServiceImpl.java
@@ -74,6 +74,8 @@ import android.annotation.CheckResult;
 import android.annotation.IntDef;
 import android.annotation.NonNull;
 import android.annotation.Nullable;
+import android.app.AlarmManager;
+import android.app.AlarmManager.OnAlarmListener;
 import android.app.AppOpsManager;
 import android.app.admin.DevicePolicyManager;
 import android.app.admin.WifiSsidPolicy;
@@ -91,6 +93,7 @@ import android.content.pm.PackageInfo;
 import android.content.pm.PackageManager;
 import android.content.pm.ResolveInfo;
 import android.content.res.Resources;
+import android.database.ContentObserver;
 import android.hardware.wifi.WifiStatusCode;
 import android.location.LocationManager;
 import android.net.DhcpInfo;
@@ -196,6 +199,7 @@ import android.os.PowerManager;
 import android.os.Process;
 import android.os.RemoteCallbackList;
 import android.os.RemoteException;
+import android.os.SystemClock;
 import android.os.UserHandle;
 import android.os.UserManager;
 import android.os.WorkSource;
@@ -326,6 +330,9 @@ public class WifiServiceImpl extends IWifiManager.Stub {
     @VisibleForTesting
     ScorerServiceConnection mScorerServiceConnection;
 
+    // Settings.Global.WIFI_OFF_TIMEOUT
+    private static final String WIFI_OFF_TIMEOUT = "wifi_off_timeout";
+
     private final ActiveModeWarden mActiveModeWarden;
     private final ScanRequestProxy mScanRequestProxy;
 
@@ -385,6 +392,17 @@ public class WifiServiceImpl extends IWifiManager.Stub {
     private final DefaultClientModeManager mDefaultClientModeManager;
 
     private final FeatureFlags mFeatureFlags;
+
+    private final OnAlarmListener mWifiTimeoutListener = new OnAlarmListener() {
+        @Override
+        public void onAlarm() {
+            if (getWifiEnabledState() == WifiManager.WIFI_STATE_ENABLED
+                    && getCurrentNetwork() == null) {
+                setWifiEnabled(mContext.getPackageName(), false);
+            }
+        }
+    };
+
     @VisibleForTesting
     public final CountryCodeTracker mCountryCodeTracker;
     private final MultiInternetManager mMultiInternetManager;
@@ -1083,6 +1101,8 @@ public class WifiServiceImpl extends IWifiManager.Stub {
             intentFilter.addAction(BluetoothAdapter.ACTION_CONNECTION_STATE_CHANGED);
             intentFilter.addAction(BluetoothAdapter.ACTION_STATE_CHANGED);
             intentFilter.addAction(PowerManager.ACTION_DEVICE_IDLE_MODE_CHANGED);
+            intentFilter.addAction(WifiManager.NETWORK_STATE_CHANGED_ACTION);
+            intentFilter.addAction(WifiManager.WIFI_STATE_CHANGED_ACTION);
             registerBroadcastReceiver(
                     new BroadcastReceiver() {
                         @Override
@@ -1130,6 +1150,9 @@ public class WifiServiceImpl extends IWifiManager.Stub {
                             } else if (PowerManager.ACTION_DEVICE_IDLE_MODE_CHANGED
                                     .equals(action)) {
                                 handleIdleModeChanged();
+                            } else if (action.equals(WifiManager.NETWORK_STATE_CHANGED_ACTION) ||
+                                    action.equals(WifiManager.WIFI_STATE_CHANGED_ACTION)) {
+                                setWifiTimeout();
                             }
                         }
                     },
@@ -1163,6 +1186,17 @@ public class WifiServiceImpl extends IWifiManager.Stub {
             mWifiInjector.getPasspointManager().handleBootCompleted();
             mWifiInjector.getInterfaceConflictManager().handleBootCompleted();
             mWifiInjector.getHalDeviceManager().handleBootCompleted();
+            mContext.getContentResolver().registerContentObserver(
+                    Settings.Global.getUriFor(WIFI_OFF_TIMEOUT),
+                    false,
+                    new ContentObserver(new Handler(mContext.getMainLooper())) {
+                        @Override
+                        public void onChange(boolean selfChange) {
+                            super.onChange(selfChange);
+                            setWifiTimeout();
+                        }
+                    }
+            );
             // HW capabilities is ready after boot completion.
             if (!mWifiGlobals.isInsecureEnterpriseConfigurationAllowed()) {
                 mWifiConfigManager.updateTrustOnFirstUseFlag(isTrustOnFirstUseSupported());
@@ -1262,6 +1296,19 @@ public class WifiServiceImpl extends IWifiManager.Stub {
         }, TAG + "#handleUserStop");
     }
 
+    private void setWifiTimeout() {
+        long wifiTimeoutMillis = Settings.Global.getLong(mContext.getContentResolver(),
+                WIFI_OFF_TIMEOUT, 0);
+        AlarmManager alarmManager = mContext.getSystemService(AlarmManager.class);
+        alarmManager.cancel(mWifiTimeoutListener);
+        if (wifiTimeoutMillis != 0) {
+            final long timeout = SystemClock.elapsedRealtime() + wifiTimeoutMillis;
+            alarmManager.setExactAndAllowWhileIdle(AlarmManager.ELAPSED_REALTIME_WAKEUP, timeout,
+                    TAG, new HandlerExecutor(new Handler(mContext.getMainLooper())), null,
+                    mWifiTimeoutListener);
+        }
+    }
+
     /**
      * See {@link android.net.wifi.WifiManager#startScan}
      *
